plugins/experimental/traffic_dump/traffic_dump.cc

```diff
@@ -0,0 +1,592 @@
+esc_json_out(const char *buf, int64_t len, std::ostream &jsonfile)
+{
+  if (buf == nullptr)
+    return 0;
+  int64_t idx, prevIdx = 0;
+  for (idx = 0; idx < len; idx++) {
+    char c = *(buf + idx);
+    switch (c) {
+    case '"':
+    case '\\': {
+      min_write(buf, prevIdx, idx, jsonfile);
+      jsonfile << "\\" << c;
+      break;
+    }
+    case '\b': {
+      min_write(buf, prevIdx, idx, jsonfile);
+      jsonfile << "\\b";
+      break;
+    }
+    case '\f': {
+      min_write(buf, prevIdx, idx, jsonfile);
+      jsonfile << "\\f";
+      break;
+    }
+    case '\n': {
+      min_write(buf, prevIdx, idx, jsonfile);
+      jsonfile << "\\n";
+      break;
+    }
+    case '\r': {
+      min_write(buf, prevIdx, idx, jsonfile);
+      jsonfile << "\\r";
+      break;
+    }
+    case '\t': {
+      min_write(buf, prevIdx, idx, jsonfile);
+      jsonfile << "\\t";
+      break;
+    }
+    default: {
+      if ('\x00' <= c && c <= '\x1f') {
+        min_write(buf, prevIdx, idx, jsonfile);
+        jsonfile << "\\u" << std::hex << std::setw(4) << std::setfill('0') << (int)c;
+      }
+      break;
+    }
+    }
+  }
+  min_write(buf, prevIdx, idx, jsonfile);
+
+  return len;
+}
+
+/// escape_json(): escape chars in a string and returns json string
+static std::string
+escape_json(std::string const &s)
+{
+  std::ostringstream o;
+  esc_json_out(s.c_str(), s.length(), o);
+  return o.str();
+}
+static std::string
+escape_json(const char *buf, int64_t size)
+{
+  std::ostringstream o;
+  esc_json_out(buf, size, o);
+  return o.str();
+}
+
+/// json_entry(): Formats to map-style entry i.e. "field": "value"
+// static inline std::string
+// json_entry(std::string const &name, std::string const &value)
+// {
+//   return "\"" + escape_json(name) + "\": \"" + escape_json(value) + "\"";
+// }
+
+static inline std::string
+json_entry(std::string const &name, const char *buf, int64_t size)
+{
+  return "\"" + escape_json(name) + "\":\"" + escape_json(buf, size) + "\"";
+}
+
+/// json_entry_array(): Formats to array-style entry i.e. ["field","value"]
+// static inline std::string
+// json_entry_array(std::string const &name, std::string const &value)
+// {
+//   return "[\"" + escape_json(name) + "\", \"" + escape_json(value) + "\"]";
+// }
+
+static inline std::string
+json_entry_array(const char *name, int name_len, const char *value, int value_len)
+{
+  return "[\"" + escape_json(name, name_len) + "\", \"" + escape_json(value, value_len) + "\"]";
+}
+
+/// Helper functions to collect txn information from TSMBuffer
+static std::string
+collect_headers(TSMBuffer &buffer, TSMLoc &hdr_loc, int64_t body_bytes)
+{
+  std::string result = "{";
+  int len            = 0;
+  const char *cp     = nullptr;
+  TSMLoc url_loc     = nullptr;
+
+  // Log scheme+method or status+reason based on header type
+  if (TSHttpHdrTypeGet(buffer, hdr_loc) == TS_HTTP_TYPE_REQUEST) {
+    // 1. "version"
+    int version = TSHttpHdrVersionGet(buffer, hdr_loc);
```

Comment:
I don't think this will capture HTTP/2.

Reply:
@dyrock clarified that the H1/H2 information is captured at the session level which seems appropriate.